File: packages/contracts/test/authority.test.mjs
import assert from "node:assert/strict";
import { createPublicKey, verify } from "node:crypto";
import { readFile } from "node:fs/promises";
import test from "node:test";
import Ajv2020 from "ajv/dist/2020.js";
import { authorityProofTranscript } from "../src/authority-proof.mjs";

test("Authority foundation accepts explicit references and denies scope or Peer fallback", async () => {
  const schema = JSON.parse(await readFile(new URL("../schemas/authority/foundation.schema.json", import.meta.url)));
  const cases = JSON.parse(await readFile(new URL("./fixtures/authority.json", import.meta.url)));
  const validator = new Ajv2020({ strict: true }).addSchema(schema);
  for (const item of cases) assert.equal(validator.getSchema(`${schema.$id}#/$defs/${item.kind}`)(item.value), item.valid, item.description);
});

test("Authority proof transcript and Ed25519 vector match the Go fixture", async () => {
  const vector = JSON.parse(await readFile(new URL("./fixtures/authority-proof-vector.json", import.meta.url)));
  const transcript = authorityProofTranscript(vector.payload);
  assert.equal(Buffer.from(transcript).toString(), vector.transcript);
  const publicKey = createPublicKey({ key: Buffer.concat([Buffer.from("302a300506032b6570032100", "hex"), Buffer.from(vector.payload.publicKey, "base64url")]), format: "der", type: "spki" });
  assert.ok(verify(null, transcript, publicKey, Buffer.from(vector.payload.signature, "base64url")));
  for (const field of ["authorityNodeId", "publicKey", "teamId", "deviceId", "ownerMemberId", "browserOrigin", "nonce", "issuedAt", "expiresAt"]) {
    const changed = authorityProofTranscript({ ...vector.payload, [field]: `${vector.payload[field]}x` });
    assert.equal(verify(null, changed, publicKey, Buffer.from(vector.payload.signature, "base64url")), false, field);
  }
});

File: packages/contracts/test/go/authority_test.go
package contracts_test

import (
	authority "convenewire.dev/contracts/generated/go/authority"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"os"
	"testing"
)

func TestAuthorityFoundationSharedFixtures(t *testing.T) {
	source, err := os.ReadFile("../fixtures/authority.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Kind        string
		Valid       bool
		Value       json.RawMessage
		Description string
	}
	if err := json.Unmarshal(source, &cases); err != nil {
		t.Fatal(err)
	}
	for _, item := range cases {
		t.Run(item.Description, func(t *testing.T) {
			var result any
			if err := authority.Decode(item.Kind, item.Value, &result); (err == nil) != item.Valid {
				t.Fatalf("validity mismatch: %v", err)
			}
		})
	}
}

func TestAuthorityProofTranscriptAndSignature(t *testing.T) {
	source, err := os.ReadFile("../fixtures/authority-proof-vector.json")
	if err != nil {
		t.Fatal(err)
	}
	var vector struct {
		Payload    authority.AuthorityProofPayload
		Transcript string
	}
	if err := json.Unmarshal(source, &vector); err != nil {
		t.Fatal(err)
	}
	transcript := authority.ProofTranscript(vector.Payload)
	if string(transcript) != vector.Transcript {
		t.Fatal("transcript differs from TypeScript")
	}
	key, err := base64.RawURLEncoding.DecodeString(vector.Payload.PublicKey)
	if err != nil {
		t.Fatal(err)
	}
	signature, err := base64.RawURLEncoding.DecodeString(vector.Payload.Signature)
	if err != nil {
		t.Fatal(err)
	}
	if !ed25519.Verify(key, transcript, signature) {
		t.Fatal("shared signature is invalid")
	}
	for _, field := range []*string{&vector.Payload.AuthorityNodeID, &vector.Payload.PublicKey, &vector.Payload.TeamID, &vector.Payload.DeviceID, &vector.Payload.OwnerMemberID, &vector.Payload.BrowserOrigin, &vector.Payload.Nonce, &vector.Payload.IssuedAt, &vector.Payload.ExpiresAt} {
		original := *field
		*field += "x"
		if ed25519.Verify(key, authority.ProofTranscript(vector.Payload), signature) {
			t.Fatal("changed scope verified")
		}
		*field = original
	}
}
